File: figures/08-graphs/snippet-dijkstra.java
    public static Result dijkstra(Graph g, int s) {
        for (long[] e : g.arcs()) if (e[2] < 0) throw new IllegalArgumentException("negative edge");
        int n = g.vertices(); long[] d = new long[n]; int[] p = new int[n];
        boolean[] finalized = new boolean[n];
        Arrays.fill(d, INF); Arrays.fill(p, -1); d[s] = 0;
        PriorityQueue<long[]> q = new PriorityQueue<>(
            Comparator.<long[]>comparingLong(x -> x[0]).thenComparingLong(x -> x[1]));
        q.add(new long[]{0, s});
        while (!q.isEmpty()) {
            long[] x = q.remove(); int u = (int)x[1];
            if (x[0] != d[u]) continue;
            if (finalized[u]) continue;
            finalized[u] = true;
            for (Graph.Edge e : g.edgesFrom(u)) {
                if (finalized[e.to()]) continue;
                long cand = safeAdd(d[u], e.weight());
                if (cand < d[e.to()]) {
                    d[e.to()] = cand; p[e.to()] = u; q.add(new long[]{cand, e.to()});
                }
            }
        }
        return new Result(d, p, false);
    }

File: figures/08-graphs/snippet-dijkstra.py
def dijkstra(g, s):
    for _, _, w in g.arcs():
        if w < 0:
            raise ValueError("negative edge")
    n = g.vertices()
    dist = [INF] * n
    parent = [-1] * n
    finalized = [False] * n
    dist[s] = 0
    pq = [(0, s)]
    while pq:
        d, u = heapq.heappop(pq)
        if d != dist[u] or finalized[u]:
            continue
        finalized[u] = True
        for v, w in g.edges_from(u):
            if finalized[v]:
                continue
            cand = dist[u] + w
            if cand < dist[v]:
                dist[v] = cand
                parent[v] = u
                heapq.heappush(pq, (cand, v))
    return dist, parent
